cli/internal/cacheitem/restore_symlink.go

```diff
@@ -0,0 +1,199 @@
+package cacheitem
+
+import (
+	"archive/tar"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/pyr-sh/dag"
+	"github.com/vercel/turborepo/cli/internal/turbopath"
+)
+
+// restoreSymlink restores a symlink and errors if the target is missing.
+func restoreSymlink(anchor turbopath.AbsoluteSystemPath, header *tar.Header, reader *tar.Reader) (turbopath.AnchoredSystemPath, error) {
+	// We don't know _anything_ about linkname. It could be any of:
+	//
+	// - Absolute Unix Path
+	// - Absolute Windows Path
+	// - Relative Unix Path
+	// - Relative Windows Path
+	//
+	// We also can't _truly_ distinguish if the path is Unix or Windows.
+	// Take for example: `/Users/turbobot/weird-filenames/\foo\/lol`
+	// It is a valid file on Unix, but if we do slash conversion it breaks.
+	// Or `i\am\a\normal\unix\file\but\super\nested\on\windows`.
+	//
+	// We also can't safely assume that paths in link targets on one platform
+	// should be treated as targets for that platform. The author may be
+	// generating an artifact that should work on Windows on a Unix device.
+	//
+	// Given all of that, our best option is to restore link targets _verbatim_.
+	// No modification, no slash conversion.
+	processedName, canonicalizeNameErr := canonicalizeName(header.Name)
+	if canonicalizeNameErr != nil {
+		return "", canonicalizeNameErr
+	}
+
+	processedLinkname := canonicalizeLinkname(anchor, processedName, header.Linkname)
+
+	// Check to see if the target exists.
+	if _, err := os.Lstat(processedLinkname); err != nil {
+		return "", errMissingSymlinkTarget
+	}
+
+	// Create the symlink.
+	// Explicitly uses the _original_ header.Linkname as the target.
+	symlinkFrom := processedName.RestoreAnchor(anchor)
+	symlinkErr := symlinkFrom.Symlink(header.Linkname)
+	if symlinkErr != nil {
+		return "", symlinkErr
+	}
+
+	// Darwin allows you to change the permissions of a symlink.
+	lchmodErr := symlinkFrom.Lchmod(fs.FileMode(header.Mode))
+	if lchmodErr != nil {
+		return "", lchmodErr
+	}
+
+	return processedName, nil
+}
+
+// restoreSymlinkMissingTarget restores a symlink and does not error if the target is missing.
+func restoreSymlinkMissingTarget(anchor turbopath.AbsoluteSystemPath, header *tar.Header, reader *tar.Reader) (turbopath.AnchoredSystemPath, error) {
+	processedName, canonicalizeNameErr := canonicalizeName(header.Name)
+	if canonicalizeNameErr != nil {
+		return "", canonicalizeNameErr
+	}
+
+	// Create the symlink.
+	// This does not support file names with `\` in them in a cross-platform manner.
+	symlinkFrom := processedName.RestoreAnchor(anchor)
+
+	// Windows:
+	// - This converts `/` into `\`.
+	// - There is no way to create a link target to something non-existent.
```

Comment:
I don't think this is correct? I thought it was just that symlinks to things that eventually turn out to be directories won't work?

[mklink](https://www.howtogeek.com/howto/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/) will let you create broken symlinks on windows

Reply:
Yup, overstated. Will correct. My tests were all of the directory issue and so wires crossed.